### PR TITLE
[operator] Introduce `gardenerscheduler` Golang component

### PR DESCRIPTION
**Pull request by oliver-goetz**
**How to categorize this PR?**

/area usability
/kind enhancement

**What this PR does / why we need it**:
Similar to https://github.com/gardener/gardener/pull/8215, this PR introduces the Golang component for the `gardener-scheduler`.
API changes and component instantiation will follow in separate PRs.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/7016

**Special notes for your reviewer**:
/cc @rfranzke @ScheererJ @timuthy 

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: docs/development/priority-classes.md [hunk 1]
```diff
@@ -46,7 +46,7 @@ When using the `gardener-operator` for managing the garden runtime and virtual c
 | Name                  | Priority  | Associated Components (Examples)                                                                                                                                                       |
 |-----------------------|-----------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | `gardener-system-500` | 999998500 | `etcd-events`, `etcd-main`, `kube-apiserver`                                                                                                                                           |
-| `gardener-system-400` | 999998400 | `gardener-resource-manager`                                                                                                                                                            |
+| `gardener-system-400` | 999998400 | `gardener-resource-manager`,                                                                                                                                                            |
```
Comment (ScheererJ) [inline docs/development/priority-classes.md]: I suppose that this change was not intentional. Can you please revert it?

**Review by ScheererJ — COMMENTED** [review]
Thanks for the change. I just have a small nit. Otherwise it looks good.

File: pkg/component/gardenerscheduler/gardener_scheduler_test.go [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+					},
+				})).To(Succeed())
+
+				Expect(fakeClient.Create(ctx, &resourcesv1alpha1.ManagedResource{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:       managedResourceNameVirtual,
+						Namespace:  namespace,
+						Generation: 1,
+					},
+					Status: healthyManagedResourceStatus,
```
Comment (plkokanov) [inline pkg/component/gardenerscheduler/gardener_scheduler_test.go]: This does not include the progressing condition. Shouldn't it be included as specified in the test description?

File: pkg/component/gardenerscheduler/gardener_scheduler.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+// Values contains configuration values for the gardener-scheduler resources.
+type Values struct {
+	// ClientConnection holds values for the client connection.
+	ClientConnection ClientConnection
+	// Image defines the container image of gardener-scheduler
+	Image string
+	// LogLevel is the level/severity for the logs. Must be one of [info,debug,error].
+	LogLevel string
+	// Replicas defines the number of replicas for the deployment
```
Comment (plkokanov) [inline pkg/component/gardenerscheduler/gardener_scheduler.go]: Nit: 
```suggestion
	// Replicas defines the number of replicas for the deployment.
```

File: pkg/component/gardenerscheduler/gardener_scheduler.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+// TimeoutWaitForManagedResource is the timeout used while waiting for the ManagedResources to become healthy or
+// deleted.
+var TimeoutWaitForManagedResource = 5 * time.Minute
+
+// Values contains configuration values for the gardener-scheduler resources.
+type Values struct {
+	// ClientConnection holds values for the client connection.
+	ClientConnection ClientConnection
+	// Image defines the container image of gardener-scheduler
```
Comment (plkokanov) [inline pkg/component/gardenerscheduler/gardener_scheduler.go]: Nit: 
```suggestion
	// Image defines the container image of gardener-scheduler.
```

File: pkg/component/gardenerscheduler/poddisruptionbudget.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+func (g *gardenerScheduler) podDisruptionBudget() *policyv1.PodDisruptionBudget {
+	return &policyv1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      DeploymentName,
+			Namespace: g.namespace,
+			Labels:    GetLabels(),
+		},
+		Spec: policyv1.PodDisruptionBudgetSpec{
+			MaxUnavailable: gardenerutils.IntStrPtrFromInt(1),
```
Comment (plkokanov) [inline pkg/component/gardenerscheduler/poddisruptionbudget.go:32]: In the charts the value for `MaxUnavailable` is given by `values.global.scheduler.replicaCount - 1`: https://github.com/gardener/gardener/blob/7fbbced193d3d6b169032767456bc518bc8d2930/charts/gardener/controlplane/charts/runtime/templates/scheduler/poddisruptionbudget.yaml#L14

Shouldn't we use the same logic here?

File: pkg/component/gardenerscheduler/rbac.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+			{
+				APIGroups: []string{coordinationv1beta1.GroupName},
+				Resources: []string{
+					"leases",
+				},
+				ResourceNames: []string{
+					schedulerv1alpha1.SchedulerDefaultLockObjectName,
+				},
+				Verbs: []string{"get", "watch", "update"},
+			},
```
Comment (plkokanov) [inline pkg/component/gardenerscheduler/rbac.go:86]: In the gardener scheduler `ClusterRole` chart there are also rules for `events` and  `configmaps`: https://github.com/gardener/gardener/blob/7fbbced193d3d6b169032767456bc518bc8d2930/charts/gardener/controlplane/charts/application/templates/clusterrole-scheduler.yaml#L20-L37
Don't we need them here as well?

File: pkg/component/gardenerscheduler/service.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+			Name:      serviceName,
+			Namespace: g.namespace,
+			Labels:    GetLabels(),
+		},
+		Spec: corev1.ServiceSpec{
+			Type:     corev1.ServiceTypeClusterIP,
+			Selector: GetLabels(),
+			Ports: []corev1.ServicePort{{
+				Name:       "metrics",
+				Port:       int32(19251),
```
Comment (plkokanov) [inline pkg/component/gardenerscheduler/service.go]:
```suggestion
				Port:       int32(metricsPort),
```

File: pkg/component/gardenerscheduler/service.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+				Protocol:   corev1.ProtocolTCP,
+				TargetPort: intstr.FromInt(19251),
```
Comment (plkokanov) [inline pkg/component/gardenerscheduler/service.go]:
```suggestion
				TargetPort: intstr.FromInt(metricsPort),
```

**Review by plkokanov — CHANGES_REQUESTED** [review]
We had a joint review with @ary1992 and mostly have some nits as feedback and a few questions.

File: pkg/component/gardenerscheduler/rbac.go (hunk 6 above)
Comment (oliver-goetz) [reply pkg/component/gardenerscheduler/rbac.go:86]: I forgot the core objects 😅

File: pkg/component/gardenerscheduler/poddisruptionbudget.go (hunk 5 above)
Comment (oliver-goetz) [reply pkg/component/gardenerscheduler/poddisruptionbudget.go:32]: I'm using the [high-availability webhook of resource-manager](https://github.com/gardener/gardener/blob/master/docs/development/high-availability.md#convenient-application-of-these-rules) to set the replicaCount in the [deployment](https://github.com/oliver-goetz/gardener/blob/fecb35328b671e6f853d3b1f6aede96e7178786a/pkg/component/gardenerscheduler/deployment.go#L47).
Thus, in this component we don't know the number of replicas the webhook is going to set.

File: pkg/component/gardenerscheduler/gardener_scheduler.go [hunk 9]
```diff
@@ -42,6 +49,22 @@ var TimeoutWaitForManagedResource = 5 * time.Minute
 
 // Values contains configuration values for the gardener-scheduler resources.
 type Values struct {
+	// ClientConnection holds values for the client connection.
+	ClientConnection ClientConnection
```
Comment (rfranzke) [inline pkg/component/gardenerscheduler/gardener_scheduler.go]: I don't think we need to expose this here - let's simply set QPS=100, Burst=130. We never had to change this in the past years.

File: pkg/component/gardenerscheduler/gardener_scheduler.go [hunk 10]
```diff
@@ -42,6 +49,22 @@ var TimeoutWaitForManagedResource = 5 * time.Minute
...
+	// LogLevel is the level/severity for the logs. Must be one of [info,debug,error].
+	LogLevel string
+	// SchedulerConfiguration provides the configuration for the SeedManager admission plugin.
+	Schedulers schedulerv1alpha1.SchedulerControllerConfiguration
```
Comment (rfranzke) [inline pkg/component/gardenerscheduler/gardener_scheduler.go]: After the discussion in #8265, we don't need this either. We can simply apply some default configuration in `configmaps.go` and go with `MinimalDistance`.

File: docs/development/priority-classes.md [hunk 11]
```diff
@@ -23,7 +23,7 @@ When using the `gardener-operator` for managing the garden runtime and virtual c
 |---------------------------------- |-----------|----------------------------------------------------------------------------------------------------------------|
 | `gardener-garden-system-critical` | 999999550 | `gardener-operator`, `gardener-resource-manager`, `istio`                                                      |
 | `gardener-garden-system-500`      | 999999500 | `virtual-garden-etcd-events`, `virtual-garden-etcd-main`, `virtual-garden-kube-apiserver`, `gardener-apiserver`|
-| `gardener-garden-system-400`      | 999999400 | `virtual-garden-gardener-resource-manager`, `gardener-admission-controller`                                    |
+| `gardener-garden-system-400`      | 999999400 | `virtual-garden-gardener-resource-manager`, `gardener-admission-controller`, `gardener-scheduler`              |
```
Comment (rfranzke) [inline docs/development/priority-classes.md]: I think it should go at least one level lower - `300` should be good enough. Maybe `200` is also appropriate compared to the criticality of the other components in the `300` level. WDYT?

File: pkg/component/gardenerscheduler/deployment.go [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+	deployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      DeploymentName,
+			Namespace: g.namespace,
+			Labels: utils.MergeStringMaps(GetLabels(), map[string]string{
+				resourcesv1alpha1.HighAvailabilityConfigType: resourcesv1alpha1.HighAvailabilityConfigTypeController,
+			}),
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &g.values.Replicas,
```
Comment (rfranzke) [inline pkg/component/gardenerscheduler/deployment.go]: Do we need to configure the replicas? I guess we can always set `1` here and rely on HA webhook to do the rest. We don't have hibernation for the Gardener control plane either. Same for GAC btw. Probably you can make the change there as well.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: docs/development/priority-classes.md (hunk 11 above)
Comment (oliver-goetz) [reply docs/development/priority-classes.md]: okay 😄

File: pkg/component/gardenerscheduler/gardener_scheduler.go (hunk 9 above)
Comment (oliver-goetz) [reply pkg/component/gardenerscheduler/gardener_scheduler.go]: This is the same for GAC. Shall I remove it there too when removing the `Replicas`?

Comment (rfranzke) [reply pkg/component/gardenerscheduler/gardener_scheduler.go]: I guess it makes sense 👍🏻

**Review by rfranzke — COMMENTED** [review]
Awesome!

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**